rs/execution_environment/fuzz/src/lib.rs

```diff
@@ -59,3 +79,145 @@ pub fn fuzzer_main() {
         }
     }
 }
+
+#[cfg(target_os = "linux")]
+fn syscall_monitor<F>(name: &str, sandbox: F)
+where
+    F: Fn(),
+{
+    match unsafe { fork() } {
+        Ok(ForkResult::Child) => {
+            sandbox();
+        }
+        Ok(ForkResult::Parent { child }) => {
+            std::thread::sleep(std::time::Duration::from_secs(1));
+            let allowed_syscalls: BTreeSet<Sysno> = BTreeSet::from([
+                Sysno::mmap,
+                Sysno::mprotect,
+                Sysno::munmap,
+                Sysno::madvise,
+                Sysno::sendmsg,
+                Sysno::sigaltstack,
+                Sysno::futex,
+                Sysno::close,
+            ]);
+            loop {
+                // This code employs a manual heuristic to determine which process PID to attach to,
+                // specifically targeting the one executing the Wasm code.
+                //
+                // The child process spawns a total of 14 threads. Since PIDs are monotonically increasing
+                // and the tid are stored in a BTreeSet, they are ordered based on their creation sequence.
+                //
+                // By tracing all PIDs and analyzing the associated syscalls, we observe that the critical
+                // threads to attach to are typically among the last few, specifically [n-2] and [n-1].
+                //
+                // NOTE: If the code design changes in the future, this heuristic will need to be revisited
+                // and updated accordingly.
+                let mut children = get_children(child.into());
+                for _ in 0..1 {
+                    children.pop_last();
+                }
+                let child = children.last().unwrap();
+                trace(name, Pid::from_raw(*child), &allowed_syscalls);
+            }
+        }
+        Err(err) => {
+            panic!("{} fork() failed: {}", name, err);
+        }
+    }
+}
+
+#[cfg(not(target_os = "linux"))]
+fn syscall_monitor<F>(_name: &str, sandbox: F)
+where
+    F: Fn(),
+{
+    sandbox();
+}
+
+#[cfg(target_os = "linux")]
+fn trace(name: &str, child: Pid, allowed_syscalls: &BTreeSet<Sysno>) {
+    if let Err(err) = ptrace::attach(child) {
+        println!(
+            "ptrace: failed to attach process {}::{}: {}",
+            name, child, err
+        );
+        return;
+    }
+
+    while let Ok(result) = waitpid(child, None) {
+        match result {
+            WaitStatus::Stopped(..) => {
+                if let Err(err) = ptrace::setoptions(child, Options::all()) {
+                    panic!(
+                        "ptrace: failed to setoptions process {}::{}: {}",
+                        name, child, err
+                    );
+                }
+
+                if let Ok(regs) = ptrace::getregs(child) {
+                    let sysno = Sysno::from(regs.orig_rax as u32);
+                    if !allowed_syscalls.contains(&sysno) {
```

Comment:
Will RAX have a syscall number on "stopped" status?

Reply:
Good catch, this was a copy paste left over.